File: cpp/src/gcylon/repartition.cpp
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <numeric>
#include <cylon/util/macros.hpp>
#include <gcylon/gtable_api.hpp>
#include <gcylon/net/cudf_net_ops.hpp>

namespace gcylon {

/**
 * Divide the rows evenly among all workers
 * number of workers = row_counts.size()
 * If the rows can not be divided evenly among all workers,
 * first k workers get one extra row.
 * For example:
 *   row_counts{0, 14, 0, 0}
 *   Result: {4, 4, 3, 3}
 * @param row_counts
 * @return
 */
std::vector<int32_t> DivideRowsEvenly(const std::vector<int32_t> &row_counts) {
  auto nworkers = row_counts.size();
  auto all_rows = std::accumulate(row_counts.begin(), row_counts.end(), 0);
  int32_t rows_per_worker = all_rows / nworkers;
  int32_t workers_with_extra_row = all_rows % nworkers;
  std::vector<int32_t> even_rows(nworkers, rows_per_worker);
  std::fill(even_rows.begin(), even_rows.begin() + workers_with_extra_row, rows_per_worker + 1);
  return even_rows;
}

/**
 * determine the target worker of the first row
 * calculate the global order of the first row of that worker
 * @param target_row_counts
 * @param my_row_start
 * @return
 */
std::pair<int32_t, int32_t> FirstTarget(const std::vector<int32_t> &target_row_counts, int32_t my_row_start) {
  // target worker of my first row
  int32_t first_target_rank = -1;
  // global order of the first row of the first target worker
  int32_t first_target_start_row = 0;
  for (int i = 0; i < target_row_counts.size(); ++i) {
    if (first_target_start_row + target_row_counts[i] > my_row_start) {
      first_target_rank = i;
      break;
    } else {
      first_target_start_row += target_row_counts[i];
    }
  }

  return std::make_pair(first_target_rank, first_target_start_row);
}

/**
 * calculates indicies for all-to-all operation for a worker
 * given the current row_counts and requested rows counts
 * make sure: number_of_workers == current_row_counts.size()
 * make sure: sum_of_rows == sum_of(current_row_counts) == sum_of(target_row_counts)
 * we do not check this in the function
 * @param my_rank
 * @param sum_of_rows sum of rows in distributed tables
 * @param current_row_counts
 * @param target_row_counts
 * @return
 */
std::vector<int32_t> RowIndicesToAll(int32_t my_rank,
                                     const std::vector<int32_t> &current_row_counts,
                                     const std::vector<int32_t> &target_row_counts) {

  // global order of my current first row
  auto my_row_start = std::accumulate(current_row_counts.begin(), current_row_counts.begin() + my_rank, 0);

  auto first_target = FirstTarget(target_row_counts, my_row_start);
  int32_t first_target_rank = first_target.first;
  int32_t first_target_start_row = first_target.second;

  int32_t rows_to_first_target = target_row_counts[first_target_rank] - (my_row_start - first_target_start_row);
  if (rows_to_first_target > current_row_counts[my_rank]) {
    rows_to_first_target = current_row_counts[my_rank];
  }

  std::vector<int32_t> to_all(target_row_counts.size() + 1, 0);
  to_all[first_target_rank + 1] = rows_to_first_target;
  int remaining_rows = current_row_counts[my_rank] - rows_to_first_target;
  int next_target = first_target_rank + 1;

  while(remaining_rows > 0) {
    int32_t current_target_rows = remaining_rows > target_row_counts[next_target] ? target_row_counts[next_target] : remaining_rows;
    to_all[next_target + 1] = to_all[next_target] + current_target_rows;
    remaining_rows -= target_row_counts[next_target];
    ++next_target;
  }

  ++next_target;
  // fill the remaining target values with the last row index, if any
  if (next_target < to_all.size()) {
    std::fill(to_all.begin() + next_target, to_all.end(), current_row_counts[my_rank]);
  }

  return to_all;
}

cylon::Status Repartition(const cudf::table_view &input_tv,
                          const std::shared_ptr<cylon::CylonContext> &ctx,
                          std::unique_ptr<cudf::table> &table_out,
                          const std::vector<int32_t> &rows_per_worker){

  std::vector<int32_t> current_row_counts;
  RETURN_CYLON_STATUS_IF_FAILED(
    RowCountsAllTables(input_tv.num_rows(), ctx, current_row_counts));

  std::vector<int32_t> rows_to_all;
  if (rows_per_worker.empty()) {
    auto evenly_dist_rows = DivideRowsEvenly(current_row_counts);
    rows_to_all = RowIndicesToAll(ctx->GetRank(), current_row_counts, evenly_dist_rows);
  } else {
    int32_t sum_of_current_rows = std::accumulate(current_row_counts.begin(), current_row_counts.end(), 0);
    int32_t sum_of_target_rows = std::accumulate(rows_per_worker.begin(), rows_per_worker.end(), 0);
    if (sum_of_current_rows != sum_of_target_rows) {
      return cylon::Status(cylon::Code::ValueError,
                           "Sum of target partitions does not match the sum of current partitions.");
    }
    rows_to_all = RowIndicesToAll(ctx->GetRank(), current_row_counts, rows_per_worker);
  }

  RETURN_CYLON_STATUS_IF_FAILED(
    gcylon::net::AllToAll(input_tv, rows_to_all, ctx, table_out));

  return cylon::Status::OK();
}

}// end of namespace gcylon

File: python/pygcylon/examples/repartition.py
##
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##

'''
# running this example with 2 mpi workers (-n 2) on the local machine:
mpirun -n 2 --mca opal_cuda_support 1 \
    python python/pygcylon/examples/repartition.py

# running this example with ucx enabled:
mpirun -n 2 --mca opal_cuda_support 1 \
    --mca pml ucx --mca osc ucx \
    python python/pygcylon/examples/repartition.py

# running this example with ucx and infiniband enabled:
mpirun -n 4 --mca opal_cuda_support 1 \
    --mca pml ucx --mca osc ucx \
    --mca btl_openib_allow_ib true \
    python python/pygcylon/examples/repartition.py
'''

import cupy as cp
import pycylon as cy
import pygcylon as gcy

env: cy.CylonEnv = cy.CylonEnv(config=cy.MPIConfig(), distributed=True)
print("CylonContext Initialized: My rank: ", env.rank)


def gen_df():
    start = 100 * env.rank
    row_count = 10 + 10 * env.rank
    df = gcy.DataFrame({'first': cp.random.randint(start, start + 10 * row_count, row_count),
                       'second': cp.random.randint(start, start + 10 * row_count, row_count)})
    row_counts = df.row_counts_for_all(env)
    if env.rank == 0:
        print("initial row counts: ", row_counts)
    return df, row_counts


######################
# even repartitioning
def repart(df, new_row_counts=None):
    repartedDF = df.repartition(rows_per_partition=new_row_counts, env=env)

    print("repartitioned row counts: ", repartedDF.row_counts_for_all(env))
    out_file = "repartitioned" + str(env.rank) + ".csv"
    repartedDF.to_cudf().to_csv(out_file)
    print("has written the repartitioned DataFrame to the file:", out_file)
    return repartedDF


######################
# generate partition sizes randomly
def gen_random_sizes(row_counts):
    avg = int(sum(row_counts) / len(row_counts))
    new_row_counts = [avg] * len(row_counts)

    import random
    from random import randrange
    # we need to generate the same random number in all workers
    random.seed(0)
    mid = int(len(new_row_counts)/2)
    for i in range(mid):
      rnd = randrange(avg)
      new_row_counts[i] += rnd
      new_row_counts[mid + i] -= rnd

    new_row_counts[len(new_row_counts) - 1] += sum(row_counts) - sum(new_row_counts)
    return new_row_counts


######################################################
df, row_counts = gen_df()

# even repartitioning
# repart(df)

# repartitioning with user given partition sizes
new_row_counts = gen_random_sizes(row_counts)

repartedDF = repart(df, new_row_counts)

env.finalize()
print("after finalize from the rank:", env.rank)
